modules/core/src/test/java/org/apache/ignite/internal/processors/cache/GridCachePartitionedClientTxMetricsTest.java

```diff
@@ -0,0 +1,194 @@
+
+            g.transactions().resetMetrics();
+        }
+    }
+
+    /**
+     *
+     */
+    public void testCommitPessimisticReadCommittedTxOnClient() {
+        doPutOnClient(PESSIMISTIC, READ_COMMITTED, true);
+    }
+
+    /**
+     *
+     */
+    public void testCommitPessimisticRepeatableReadTxOnClient() {
+        doPutOnClient(PESSIMISTIC, REPEATABLE_READ, true);
+    }
+
+    /**
+     *
+     */
+    public void testCommitPessimisticSerializableTxOnClient() {
+        doPutOnClient(PESSIMISTIC, SERIALIZABLE, true);
+    }
+
+    /**
+     *
+     */
+    public void testCommitOptimisticReadCommittedTxOnClient() {
+        doPutOnClient(OPTIMISTIC, READ_COMMITTED, true);
+    }
+
+    /**
+     *
+     */
+    public void testCommitOptimisticRepeatableReadTxOnClient() {
+        doPutOnClient(OPTIMISTIC, REPEATABLE_READ, true);
+    }
+
+    /**
+     *
+     */
+    public void testCommitOptimisticSerializableTxOnClient() {
+        doPutOnClient(OPTIMISTIC, SERIALIZABLE, true);
+    }
+
+    /**
+     *
+     */
+    public void testRollbackPessimisticReadCommittedTxOnClient() {
+        doPutOnClient(PESSIMISTIC, READ_COMMITTED, false);
+    }
+
+    /**
+     *
+     */
+    public void testRollbackPessimisticRepeatableReadTxOnClient() {
+        doPutOnClient(PESSIMISTIC, REPEATABLE_READ, false);
+    }
+
+    /**
+     *
+     */
+    public void testRollbackPessimisticSerializableTxOnClient() {
+        doPutOnClient(PESSIMISTIC, SERIALIZABLE, false);
+    }
+
+    /**
+     *
+     */
+    public void testRollbackOptimisticReadCommittedTxOnClient() {
+        doPutOnClient(OPTIMISTIC, READ_COMMITTED, false);
+    }
+
+    /**
+     *
+     */
+    public void testRollbackOptimisticRepeatableReadTxOnClient() {
+        doPutOnClient(OPTIMISTIC, REPEATABLE_READ, false);
+    }
+
+    /**
+     *
+     */
+    public void testRollbackOptimisticSerializableTxOnClient() {
+        doPutOnClient(OPTIMISTIC, SERIALIZABLE, false);
+    }
+
+    /**
+     * @param transactionConcurrency transaction concurrency.
+     * @param transactionIsolation transaction isolation.
+     * @param commit {@code True} if should commit, otherwise - rollback transaction.
+     */
+    private void doPutOnClient(TransactionConcurrency transactionConcurrency,
+        TransactionIsolation transactionIsolation, boolean commit) {
+        Ignite client = ignite(gridCount() - 1);
+
+        IgniteCache<Integer, String> cache = client.cache(DEFAULT_CACHE_NAME);
+
+        IgniteTransactions transactions = client.transactions();
+
+        for (int i = 0; i < EXPECTED_TXS_FINISH_COUNT; i++) {
+            try (Transaction tx = transactions.txStart(transactionConcurrency, transactionIsolation)) {
+                cache.put(i, "value");
+
+                if (commit)
+                    tx.commit();
+                else
+                    tx.rollback();
+            }
+        }
+
+        try {
+            Thread.sleep(2 * METRICS_UPDATE_FREQUENCY);
```

Comment:
May be wait for recieving metrics message instead of sleep?

Reply:
I think GridTestUtils.waitForCondition fits better